### PR TITLE
Second API rewrite to improve caching and explanations

### PR DESCRIPTION
**Pull request by juanibiapina**
This is a better version of #17 . It also invalidates #18 and #19 

## Features

- Simplify API (the only top level concept is Calculator)
- Caching of all intermediate values (Implements #4)
- Auto generate recursive explanation (Implements #5)
- Support for nested context as objects (Implements #8)
- Include input params in explanation tree
- Reuse one calculator on different branches of the context (if their schema match) (Implements #10)
- Add extra values to the explanation

Example API:

```ruby
param :base_value # simple param from context

param :optional, default: 10 # optional param with default value

param :origin do 
  param :some_value # nested param from context
end

component :price_per_distance, PricePerDistanceCalculator # normal calculator

component :origin_factor, StopFactorCalculator, :origin # sub context
component :destination_factor, StopFactorCalculator, :destination # sub context

transient :capacity do # transient factor that appears on the explanation
  # fetch database capacity or do any intermediate calculation
end
```

## Checklist

- [x] Make Calculator the only top level API
- [x] Replace composed_by with `component` and `param`
- [x] Generate explanation from composed parts
- [x] Add context to the explanation
- [x] Cache calculation values
- [x] Make explanation recursive
- [x] Add support for nested hashes in `param`
- [x] Add support for `transient` values
- [x] Allow component to receive only a part of the context
- [x] Allow default values for params

### CODE REVIEW

**Comment by marceloboeira** [comment]
@juanibiapina ready to merge? I haven't seen anything that worries me. I guess I will just recreate the "uber example" for the sake of demonstration/documentation but other than that I am okay.

File: spec/prezzo/composable_spec.rb [hunk 1]
```diff
@@ -59,6 +49,16 @@ def calculate
     end
   end
 
+  describe "params" do
+    it "declares methods for the params" do
+      expect(calculator.methods).to include(:some_param)
+    end
+
+    it "reads the param from the context" do
+      expect(calculator.some_param).to eq(42.3)
```
Comment (juanibiapina) [inline spec/prezzo/composable_spec.rb]: probably not, but I haven't changed this behavior

File: lib/prezzo/explainable.rb [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
 
-    module ClassMethods
-      def explain_with(*options)
-        define_method(:explain) do
-          options.each_with_object({}) do |method, explanation|
-            explanation[method] = send(method)
-          end
-        end
+      components = cached_components.reduce({}) do |acc, (name, component)|
+        acc[name] = component.calculate
```
Comment (juanibiapina) [inline lib/prezzo/explainable.rb]: this is done in a more recent commit, if I understand correctly, right?

**Comment by juanibiapina** [comment]
Maybe make it a release candidate for releasing?

**Comment by marceloboeira** [comment]
@juanibiapina release from the branch?

**Comment by juanibiapina** [comment]
Merge and release I think...

**Pull request merged**